### PR TITLE
Adding static analysis topic from wiki

### PR DESCRIPTION
**Pull request by dobooth**
## Purpose of this pull request

This pull request (PR) adds a document about setting up static analysis that was going private to do magento-commerce changes. Wanted to keep the info available.

## Affected DevDocs pages

https://devdocs.magento.com/guides/v2.4/test/static/static-analysis.html

whatsnew
Added the [Setting up static analysis](https://devdocs.magento.com/guides/v2.4/test/static/static-analysis.html) topic.

### CODE REVIEW

File: src/guides/v2.3/test/static/static-analysis.md [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+  - test
+---
+
+## Introduction
+
+This document provides setup steps to get most static analyzers that are used in our build pipeline working locally in PHPStorm.  This guide is written specifically for those using PHPStorm on MacOS.
+
+## Before you begin
+
+If you haven't already, make sure that node and npm are installed. Then, in the Magento root directory, run `npm install`.  After installation, ensure there is a `node_modules` directory in the root of your project before proceeding.
```
Comment (BarnyShergold) [inline src/guides/v2.3/test/static/static-analysis.md:15]:
```suggestion
If you haven't already, make sure that Node and NPM are installed. Then, in the Magento root directory, run `npm install`.  After installation, ensure there is a `node_modules` directory in the root of your project before proceeding.
```

File: src/guides/v2.3/test/static/static-analysis.md [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+1. Click the "Manual ESLint configuration" radio button.
+1. Fill in the adjacent input fields with the path to your node binary (the result of outputting `which node` in your terminal).
+1. Enter the path to your ESLint package: `[magento_root]/node_modules/eslint`
+1. Click the "Configuration File" radio button and in the adjacent input field enter the path to the ESLint file Magento uses, which is located in `dev/tests/static/testsuite/Magento/Test/Js/_files/eslint/.eslintrc-magento`.
+
+See the image below for example configuration:
+
+![ESLint]({{site.baseurl}}/common/images/static-eslint.png)
+
+To verify it works, in any JS file add `/** Hello world */` as a doc comment to any method you should see warning about the comment being on one line
```
Comment (BarnyShergold) [inline src/guides/v2.3/test/static/static-analysis.md:47]:
```suggestion
To verify it works, in any JS file add `/** Hello world */` as a doc comment to any method, you should see a warning about the comment being on one line.
```

File: src/guides/v2.3/test/static/static-analysis.md [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+1. Go to PHPStorm preferences > Languages & Frameworks > PHP > Quality Tools > PHP_CodeSniffer.
+1. Click the `...` button to bring up another configuration modal.
+1. Enter the PHP_CodeSniffer path: `[magento_root]/vendor/bin/phpcs`
+
+### Configuring for Magento Coding Standard
+
+1. Go to the [Magento Coding Standard GitHub Repository](https://github.com/magento/magento-coding-standard)
+1. Follow instructions within the README to install the Magento Coding Standard for PHPCS.  Verify it is installed with `vendor/bin/phpcs -i`.  You should see `Magento2` in the output.
+1. Go to PHPStorm preferences > Editor > Inspections, and in the adjacent window go to PHP > Quality Tools > PHP_CodeSniffer validation.
+1. Under Coding Standard dropdown, select Magento2.
```
Comment (BarnyShergold) [inline src/guides/v2.3/test/static/static-analysis.md:60]:
```suggestion
1. Under Coding Standard dropdown, select `Magento2`.
```

File: src/guides/v2.3/test/static/static-analysis.md [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+![PHPCS]({{site.baseurl}}/common/images/static-codesniff.png)
+
+To verify it works, add the following PHP snippet: `$base = basename($_SERVER['SCRIPT_FILENAME']);`.  It should give you a warning about use of `basename` being forbidden as well as use of superglobals.
+
+## PHPMD
+
+1. Go to PHPStorm preferences > Languages & Frameworks > PHP > Quality Tools > Mess Detector.
+1. Click the `...` button to bring up another configuration modal.
+1. Enter the PHP Mess Detector path: `[magento_root]/vendor/phpmd/phpmd/src/bin/phpmd`
+1. Go to PHPStorm preferences > Editor > Inspections, and in the adjacent window go to  > PHP > Quality Tools > PHP Mess Detector validation.
```
Comment (BarnyShergold) [inline src/guides/v2.3/test/static/static-analysis.md:73]:
```suggestion
1. Go to PHPStorm preferences > Editor > Inspections and in the adjacent window go to  > PHP > Quality Tools > PHP Mess Detector validation.
```

**Review by BarnyShergold — CHANGES_REQUESTED** [review]
This is an excellent new Topic. The instructions are very clear and concise. I checked a few locations for the settings and the instructions were spot on. I am unable to fully check all the technical accuracies as I don't have some of the features installed on my MAC (I am not a developer) 

A second review by a Technical Architect for might be recommended.

File: src/guides/v2.3/test/static/static-analysis.md (hunk 1 above)
Comment (dobooth) [reply src/guides/v2.3/test/static/static-analysis.md:15]: Those are generally accepted as lowercase.

Comment (hguthrie) [inline src/guides/v2.3/test/static/static-analysis.md:15]: No contractions, and maybe can emphasize node and npm with code style.

```suggestion
If you have not already, verify that `node` and `npm` are installed. Then, in the Magento root directory, run `npm install`.  After installation, ensure there is a `node_modules` directory in the root of your project before proceeding.
```

Comment (BarnyShergold) [reply src/guides/v2.3/test/static/static-analysis.md:15]: In that case (please add @hguthrie's suggestion for emphasis)

**Review by BarnyShergold — CHANGES_REQUESTED** [review]

File: src/guides/v2.3/test/static/static-analysis.md [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+1. Go to PHPStorm preferences > Languages & Frameworks > PHP > Quality Tools > PHP_CodeSniffer.
+1. Click the `...` button to bring up another configuration modal.
+1. Enter the PHP_CodeSniffer path: `[magento_root]/vendor/bin/phpcs`
+
+### Configuring for Magento Coding Standard
+
+1. Go to the [Magento Coding Standard GitHub Repository](https://github.com/magento/magento-coding-standard)
+1. Follow instructions within the README to install the Magento Coding Standard for PHPCS.  Verify it is installed with `vendor/bin/phpcs -i`.  You should see `Magento2` in the output.
+1. Go to PHPStorm preferences > Editor > Inspections, and in the adjacent window go to PHP > Quality Tools > PHP_CodeSniffer validation.
+1. Under Coding Standard dropdown, select `Magento2`.
```
Comment (hguthrie) [inline src/guides/v2.3/test/static/static-analysis.md:60]:
```suggestion
1. Under **Coding Standard** dropdown, select `Magento2`.
```

File: src/guides/v2.3/test/static/static-analysis.md [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+
+See the image below for example configuration:
+
+![PHPCS]({{site.baseurl}}/common/images/static-codesniff.png)
+
+To verify it works, add the following PHP snippet: `$base = basename($_SERVER['SCRIPT_FILENAME']);`.  It should give you a warning about use of `basename` being forbidden as well as use of superglobals.
```
Comment (hguthrie) [inline src/guides/v2.3/test/static/static-analysis.md:66]: Changing wording for consistency.

```suggestion
To verify it works, add the following PHP snippet: `$base = basename($_SERVER['SCRIPT_FILENAME']);`.  You should see a warning that the use of `basename` is forbidden, as well as the use of _superglobals_.
```

File: src/guides/v2.3/test/static/static-analysis.md [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+1. Click the "Manual ESLint configuration" radio button.
+1. Fill in the adjacent input fields with the path to your node binary (the result of outputting `which node` in your terminal).
+1. Enter the path to your ESLint package: `[magento_root]/node_modules/eslint`
+1. Click the "Configuration File" radio button and in the adjacent input field enter the path to the ESLint file Magento uses, which is located in `dev/tests/static/testsuite/Magento/Test/Js/_files/eslint/.eslintrc-magento`.
+
+See the image below for example configuration:
+
+![ESLint]({{site.baseurl}}/common/images/static-eslint.png)
+
+To verify it works, in any JS file add `/** Hello world */` as a doc comment to any method, you should see a warning about the comment being on one line.
```
Comment (hguthrie) [inline src/guides/v2.3/test/static/static-analysis.md:47]:
```suggestion
To verify it works, in any JS file add `/** Hello world */` as a doc comment to any method, and you should see a warning about the comment being on one line.
```

File: src/guides/v2.3/test/static/static-analysis.md [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+1. Click the `...` button to bring up another configuration modal.
+1. Enter the PHP Mess Detector path: `[magento_root]/vendor/phpmd/phpmd/src/bin/phpmd`
+1. Go to PHPStorm preferences > Editor > Inspections and in the adjacent window go to  > PHP > Quality Tools > PHP Mess Detector validation.
+1. Under custom rulesets, add the path to the PHPMD ruleset Magento uses: `[magento_root]/dev/tests/static/testsuite/Magento/Test/Php/_files/phpmd/ruleset.xml`
+
+See the image below for example configuration:
+
+![PHPMD]({{site.baseurl}}/common/images/static-md.png)
+
+To verify it works, add an unused private method to a class and you should see a warning from PHPMD about it not being used.
```
Comment (hguthrie) [inline src/guides/v2.3/test/static/static-analysis.md:80]:
```suggestion
To verify it works, add an unused private method to a class, and you should see a warning from PHPMD about it not being used.
```

**Review by hguthrie — CHANGES_REQUESTED** [review]
YAY! Good instructions. 🏆 

Made a few suggestions for consistency and for style, but there was nothing egregious. I made a guess on a few things, so double check.

It is _optional_, but you can use the procedure title convention to state what the steps do. "To install blah blah:" for example. You can find that usage in the style guide.

File: src/guides/v2.3/test/static/static-analysis.md [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+If you have not already, verify that `node` and `npm` are installed. Then, in the Magento root directory, run `npm install`.  After installation, ensure there is a `node_modules` directory in the root of your project before proceeding.
+
+{:bs-callout-info}
+For all of the static configuration installations below involving the PHPStorm Preferences dialog, make sure you are clicking the "Apply" button before clicking "OK" to close the Preferences dialog.
+
+## Javascript Code Style check
+
+{:.bs-callout-info}
+JSCS is deprecated in PHPStorm 2020.2. It must be installed via a plugin.
+
```
[the remark this answers is not in the document]
Comment (dobooth) [reply src/guides/v2.3/test/static/static-analysis.md:24]: Hmm...not familiar with the procedure title.

File: src/guides/v2.3/test/static/static-analysis.md [hunk 10]
```diff
@@ -0,0 +1,81 @@
...
+{:bs-callout-info}
+For all of the static configuration installations below involving the PHPStorm Preferences dialog, make sure you are clicking the "Apply" button before clicking "OK" to close the Preferences dialog.
+
+## Javascript Code Style check
+
+{:.bs-callout-info}
+JSCS is deprecated in PHPStorm 2020.2. You must install the JSCS plugin.
+{:.procedure}
+To install the JSCS plugin:
+1. Go to PHPStorm preferences > Languages & Frameworks > JavaScript > Code Quality Tools > JSCS.
```
Comment (hguthrie) [inline src/guides/v2.3/test/static/static-analysis.md:26]: Just need the blank lines here.
```suggestion
JSCS is deprecated in PHPStorm 2020.2. You must install the JSCS plugin.

{:.procedure}
To install the JSCS plugin:

1. Go to PHPStorm preferences > Languages & Frameworks > JavaScript > Code Quality Tools > JSCS.
```

**Review by hguthrie — APPROVED** [review]

**Comment by dobooth** [comment]
running tests

**Pull request merged**